### ISSUE TITLE
Is std or other error wrapping a good thing ?

### ISSUE THREAD

**Issue by luqasz** [luqasz/avrisp#10]
@phaazon

I've wrapped `Io::Error` into my own `ErrorKind` enum. I don't know if it's ok in rust. My thought process was: "If I have function which can return my defined error or some other then I need to wrap them all into my `ErrorKind`"

**Comment by hadronized** [comment]
That’s correct. Error handling in Rust can be done in a wide variety of ways. Something that I personally like to do is to use one error type for a given _scope of responsibility_, so that I don’t have a single error type for my whole lib / application. The most extreme version of this is to have one error type by function:

```rust
fn foo() -> Result<(), FooError> {}

fn bar() -> Result<(), BarError> {}
```

If both functions work in the same unit of responsibility, you should be using the same error type, but let’s imagine for a second that both functions work on different parts of your logic so they require a different error type. The problem comes if you want to call both functions from a `quux` function, which might have its own error type, too — `QuuxError`. The way I see things and how things should, at least to me, be done, is that `QuuxError` should allow to _lift_ a `FooError` or `BarError`. You did that with your enum. A nice syntactic sugar of Rust with the `?` operator is that it works on `E: Into<ReturnErrorType>` rather than `ReturnErrorType` directly. So you can simply do:

```rust
enum QuuxError {
  FooError(FooError),
  BarError(BarError),
}

impl From<FooError> for QuuxError {
  // …
}

impl From<BarError> for QuuxError {
  // …
}

fn quux() -> Result<(), QuuxError> {
  let _ = foo()?;
  let _ = bar()?;

  Ok(())
}
```

If you don’t have that `From` implementor to convert from a nested error to its parent error, you can simply `map_err` before calling `?`. Sometimes you will have two variants accepting the same underlying error type, so you can’t have that `From` impl. Imagine:

```rust
enum QuuxError {
  CannotFindFoo(FooError),
  ErrorParsingFoo(FooError),
}
```

You can’t impl `From<FooError> for QuuxError`, because it would be ambiguous which variant to end up in. Here, use `map_err`:

fn quux() -> Result<(), QuuxError> {
  let _ = find_foo().map_err(QuuxError::CannotFindFoo)?;
  let _ = parse_foo().map_err(QuuxError::ErrorParsingFoo)?;

  Ok(())
}
```

Finally, you will find a bunch of libraries solving the error problem by basically turning everything dynamic for you (using trait objects like `dyn anyhow::Error`. I honestly don’t recommend these libraries. They are very invasive in your error workflow and they don’t really solve the problem as they just shift the problem via dynamic dispatch. However, there are some cool crates I don’t recall the names to prevent you from having to write all these `From` implementors.

**Comment by luqasz** [comment]
Thx. Now my question is. How to match to get specific error ? Let's say that function will return error enum (not  a concrete one). Calling it will result in returning error enum. I'd like to know what was the root cause. I can see that `Io::Error` has a `kind()` method, but not my errors. Do I have to use match ?

**Comment by hadronized** [comment]
The root cause depends on the enum you’re using. With the `QuuxError` above, matching on it will provide you the specific error by capturing the variant’s content.
